airbyte-integrations/connectors/source-zendesk-support/unit_tests/unit_test.py

```diff
@@ -159,26 +152,25 @@ def test_default_start_date():
     [
         (TEST_CONFIG, "aW50ZWdyYXRpb24tdGVzdEBhaXJieXRlLmlvL3Rva2VuOmFwaV90b2tlbg=="),
         (TEST_CONFIG_OAUTH, "test_access_token"),
-        (TEST_OLD_CONFIG, "aW50ZWdyYXRpb24tdGVzdEBhaXJieXRlLmlvL3Rva2VuOmFwaV90b2tlbg=="),
```

Comment:
This one is interesting. Using the latest version of the source and this config:
```
{
    "subdomain": "d3v-airbyte",
    "start_date": "2022-06-01T00:00:00Z",
    "ignore_pagination": true,
    "auth_method": {
        "auth_method": "api_token",
        "email": <redacted>,
        "api_token": <redacted>
    }
}
```

... I would get the following error during a read:
```
{
    "type": "TRACE",
    "trace": {
        "type": "ERROR",
        "emitted_at": 1731073703195,
        "error": {
            "message": "Something went wrong in the connector. See the logs for more details.",
            "internal_message": "The path from `authenticator_selection_path` is not found in the config.",
            "stack_trace": "Traceback (most recent call last):
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/auth/selective_authenticator.py\", line 30, in __new__
    selected_key = str(dpath.get(config, authenticator_selection_path))
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/dpath/__init__.py\", line 192, in get
    raise KeyError(glob)
KeyError: ['credentials', 'credentials']

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/main.py\", line 8, in <module>
    run()
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/run.py\", line 14, in run
    launch(source, sys.argv[1:])
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/entrypoint.py\", line 263, in launch
    for message in source_entrypoint.run(parsed_args):
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/entrypoint.py\", line 133, in run
    yield from map(AirbyteEntrypoint.airbyte_message_to_string, self.read(source_spec, config, config_catalog, state))
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/entrypoint.py\", line 191, in read
    for message in self.source.read(self.logger, config, catalog, state):
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/manifest_declarative_source.py\", line 165, in read
    yield from super().read(logger, config, catalog, state)
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/abstract_source.py\", line 98, in read
    stream_instances = {s.name: s for s in self.streams(config)}
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/source.py\", line 176, in streams
    declarative_streams = super().streams(args)
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/manifest_declarative_source.py\", line 95, in streams
    source_streams = [
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/manifest_declarative_source.py\", line 96, in <listcomp>
    self._constructor.create_component(
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py\", line 292, in create_component
    return self._create_component_from_model(model=declarative_component_model, config=config, **kwargs)
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py\", line 300, in _create_component_from_model
    return component_constructor(model=model, config=config, **kwargs)
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py\", line 738, in create_declarative_stream
    retriever = self._create_component_from_model(
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py\", line 300, in _create_component_from_model
    return component_constructor(model=model, config=config, **kwargs)
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py\", line 1224, in create_simple_retriever
    requester = self._create_component_from_model(model=model.requester, decoder=decoder, config=config, name=name)
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py\", line 300, in _create_component_from_model
    return component_constructor(model=model, config=config, **kwargs)
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py\", line 897, in create_http_requester
    self._create_component_from_model(model=model.authenticator, config=config, url_base=model.url_base, name=name, decoder=decoder)
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py\", line 300, in _create_component_from_model
    return component_constructor(model=model, config=config, **kwargs)
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py\", line 1186, in create_selective_authenticator
    return SelectiveAuthenticator(  # type: ignore[abstract]
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/auth/selective_authenticator.py\", line 32, in __new__
    raise ValueError(\"The path from `authenticator_selection_path` is not found in the config.\") from err
ValueError: The path from `authenticator_selection_path` is not found in the config.
",
            "failure_type": "system_error"
        }
    }
}
```

The config has been updated in our GSM on `3/15/22, 7:49 AM`. I would assume it was because of [this change](https://github.com/airbytehq/airbyte/pull/11162/files#diff-2f6a32b67d7c88a2a4ccbb67200caeb69066953dea81994ec68a171414a11d0d). However, the manifest.yaml never took into account the old config format [see first commit [here](https://github.com/airbytehq/airbyte/pull/36823/files#diff-40506820d30657ee0d53226799b3489cf4cc350afe6147ee46c3efb551237bf7R15-R21)). Hence, I assume that old configs don't work since the introduction of manifest.yaml for any source that was low-code.

Why is this breaking now? It is because we call `streams` in `__init__` now which instantiate those streams in order to dispatch them as `concurrent_streams` or `sequential_streams`. We've made this conscious decision with @brianjlai knowing the consequences because else, the variable `concurrent_streams` and `sequential_streams` would have been stateful (sometimes instantiated, some other times not) which would have had complexity in `ConcurrentDeclarativeSource`

Reply:
Thanks for the detailed writeup, i think we can just accept that if a connector is still using a 2 1/2 year old config, we don't have to guarantee it's going to work and we can accept this type of failure might exist